app/app.go

```diff
@@ -602,12 +612,114 @@ func (app *RegenApp) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) a
 // EndBlocker application updates every end block
 func (app *RegenApp) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
 	resp := app.mm.EndBlock(ctx, req)
-	events, vals := app.smm.EndBlock(ctx, req)
-	if len(resp.ValidatorUpdates) > 0 && len(vals) > 0 {
-		panic("validator EndBlock updates already set by the SDK Module Manager")
+
+	// at the upgrade height
+	// for all validators in app state (app.StakingKeeper.GetAllValidators), create a ValidatorUpdate with the correct power.
+	// for all remaining validators in []VoteInfo that are not in app state, set their voting power to zero.
+	// reset jailed validator missed blocks counter so that the validator won't be immediately slashed for downtime after unjail.
+
+	// after the upgrade height continue the correct behavior.
+	if app.runEndBlockerUpgrade {
+		app.runEndBlockerUpgrade = false
+
+		// TODO: update with mainnet validatorset
+		// replace with
+		// validatorsPubkeys, err := GetMainnetValidatorSetPubkeys()
+		validatorsPubkeys, err := GetTestnetValidatorSetPubKeys()
```

Comment:
@aleem1314 can we swap this out and set to `GetTestnetValidatorSetPubKeys()` or `GetMainnetValidatorSetPubKeys()` depending on chain-id == `regen-1` or `regen-redwood-1` ?

Reply:
updated